### PR TITLE
TYPE: Properly pass variables into extracted function

### LINKED ISSUE #2188: Properly pass variables into extracted function
We always pass variables by value and it can lead to incorrect code.
We should correctly determine how to pass variable: by value or by reference.

```rust
fn foo() {
    let vec = vec![1, 2, 3];
    <selection>println!("{}", vec.len());</selection>
    println!("{}", vec[0]);
}
```

**Expected:**
```rust
fn foo() {
    let vec = vec![1, 2, 3];
    bar(&vec);
    println!("{}", vec[0]);
}

fn bar(vec: &Vec<i32>) {
    println!("{}", vec.len());
}
```

**Actual:**
```rust
fn foo() {
    let vec = vec![1, 2, 3];
    bar(vec);
    println!("{}", vec[0]); // error[E0382]: use of moved value: `vec`
}

fn bar(vec: Vec<i32>) {
    println!("{}", vec.len()); 
}
```

### PR DESCRIPTION
**Pull request by t-kameyama**
Fixes #2188

### CODE REVIEW

**Comment by Undin** [comment]
Rebase onto master, please)

File: src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt [hunk 1]
```diff
@@ -150,12 +155,14 @@ class RsExtractFunctionConfig private constructor(
...
             val parameters = letBindings
-                .filter { it.textOffset <= start }
-                .filter {
-                    ReferencesSearch.search(it, LocalSearchScope(fn))
-                        .any { ref -> ref.element.textOffset in start..end }
+                .mapNotNull {
+                    if (it.textOffset > start) return@mapNotNull null
+                    val result = ReferencesSearch.search(it, LocalSearchScope(fn))
+                    if (result.none { it.element.textOffset in start..end }) return@mapNotNull null
+                    val requiredBorrowing = result.any { it.element.textOffset > end }
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt]: Let's use a little bit different strategy here - pass a variable with minimal required "attribute". I mean if it is enough to pass only reference, let's pass reference even we can move variable and don't break code.
So, for 
```rust
fn foo() {
    let mut v = Vec::new();
    <selection>v.push(123);</selection>
}
```
result should be
```rust
fn foo() {
    let mut v = Vec::new();
    bar(&mut v);
}

fn bar(v: &mut Vec<i32>) {
    v.push(123);
}
```

Comment (Undin) [reply src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt]: Also current implementation can pass references by reference
```rust
fn foo() {
    let s = "str";
    <selection>println!("{}", s);</selection>
    println!("{}", s);
}
```
=>
```rust
fn foo() {
    let s = "str";
    bar(&s);
    println!("{}", s);
}

fn bar(s: &&str) {
    println!("{}", s);
}
```

File: src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt [hunk 2]
```diff
@@ -150,12 +155,14 @@ class RsExtractFunctionConfig private constructor(
...
+                        && it.type !is TyPrimitive && !ImplLookup.relativeTo(it).isCopy(it.type)
```
[the remark this answers is not in the document]
Comment (Undin) [reply src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt]: Hm, looks like that not all primitive types implement `Copy` trait (for example `str`).
And add a comment why we have this check here to drop it when `ImplLookup. isCopy` will support primitive types

File: src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt [hunk 3]
```diff
@@ -40,28 +42,31 @@ class ReturnValue(val expression: String?, val type: Ty) {
     }
 }
 
-class Parameter(val name: String, val type: Ty? = null, val mutability: Mutability = Mutability.IMMUTABLE) {
+class Parameter(val name: String, val type: Ty? = null, val mutability: Mutability = Mutability.IMMUTABLE,
```
[the remark this answers is not in the document]
Comment (t-kameyama) [reply src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt]: It seems `ImplLookup.isCopy` does not support primitive types.

File: src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (t-kameyama) [reply src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt]: It seems ImplLookup.isCopy does not support primitive types.

File: src/test/kotlin/org/rust/lang/refactoring/RsExtractFunctionTest.kt [hunk 4]
```diff
@@ -726,6 +726,76 @@ class RsExtractFunctionTest : RsTestBase() {
         false,
         "bar")
 
+        fun `test extract a function with reference parameter`() = doTest("""
```
Comment (Undin) [inline src/test/kotlin/org/rust/lang/refactoring/RsExtractFunctionTest.kt]: Let's split this test into several small tests to check one case per tests
I think it's more convenient to check what cases are tested

File: src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt [hunk 5]
```diff
@@ -40,28 +39,36 @@ class ReturnValue(val expression: String?, val type: Ty) {
...
 
-class Parameter(val name: String, val type: Ty? = null, val mutability: Mutability = Mutability.IMMUTABLE) {
+class Parameter(
+    val name: String,
+    val type: Ty? = null,
+    val reference: Reference = Reference.NONE
+) {
+
+    enum class Reference(val text: String) {
+        MUTABLE("&mut "), BORROWING("& "), NONE("")
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt]: What about passing a variable by mutable value?
```rust
fn foo(mut v: Vec<i32>) {}

fn main() {
    let mut v = Vec::new();
    <selection>v.push(1);
    foo(v);</selection>
}
```
should be
```rust
fn foo(mut v: Vec<i32>) {}

fn main() {
    let mut v = Vec::new();
    bar(v);
}

fn bar(mut v: Vec<i32>) {
    v.push(1);
    foo(v);
}

```
but current version generates the following
```rust
fn foo(mut v: Vec<i32>) {}

fn main() {
    let mut v = Vec::new();
    bar(&mut v);
}

fn bar(v: &mut Vec<i32>) {
    v.push(1);
    foo(v);
}

```

**Comment by t-kameyama** [comment]
Fixed

File: src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt [hunk 6]
```diff
@@ -149,13 +157,28 @@ class RsExtractFunctionConfig private constructor(
...
-                .filter { it.textOffset <= start }
-                .filter {
-                    ReferencesSearch.search(it, LocalSearchScope(fn))
-                        .any { ref -> ref.element.textOffset in start..end }
+                .mapNotNull {
+                    if (it.textOffset > start) return@mapNotNull null
+                    val result = ReferencesSearch.search(it, LocalSearchScope(fn))
+                    if (result.none { it.element.textOffset in start..end }) return@mapNotNull null
+
+                    val requiredBorrowing = result.any { it.element.textOffset > end }
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt]: `result.any { it.element.textOffset > end }` condition is not enough here
In the following case we still can pass variable by reference
```rust
fn foo() {
    let a = Vec::new();
    <selection>f(&a);</selection>
}

fn f(a: &Vec<i32>) {}
```
but pass it by value
```rust
fn foo() {
    let a = Vec::new();
    bar(a);
}

fn bar(a: Vec<i32>) {
    f(&a);
}

fn f(a: &Vec<i32>) {}
```

File: src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt [hunk 7]
```diff
@@ -149,13 +157,28 @@ class RsExtractFunctionConfig private constructor(
...
+                        && it.type !== TyBool && it.type !== TyChar && it.type !is TyInteger && it.type !is TyFloat
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt]: Let's add a comment here to explain why we manually check primitive types

Comment (Undin) [reply src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt]: And looks like there isn't difference between `!==` and `!=` here

File: src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt [hunk 8]
```diff
@@ -40,28 +36,40 @@ class ReturnValue(val expression: String?, val type: Ty) {
...
+    enum class Reference(val text: String) {
+        MUTABLE("&mut "), IMMUTABLE("& "), NONE("")
+    }
+
     companion object {
-        fun direct(value: RsPatBinding): Parameter {
-            return Parameter(value.referenceName, value.type, value.mutability)
+        fun direct(value: RsPatBinding, requiredBorrowing: Boolean, requiredMutableValue: Boolean): Parameter {
+            val reference = when {
+                requiredMutableValue -> Reference.NONE
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt]:
```rust
fn foo(mut a: Vec<i32>) {
    <selection>f(&mut a);</selection>
    f(&mut a);
}

fn f(a: &mut Vec<i32>) {}
```
=>
```rust
fn foo(mut a: Vec<i32>) {
    bar(a);
    f(&mut a); // error[E0382]: use of moved value: `a`
}

fn bar(mut a: Vec<i32>) {
    f(&mut a);
}

fn f(a: &mut Vec<i32>) {}
```

**Comment by t-kameyama** [comment]
Fixed

**Comment by Undin** [comment]
Sorry for so slow response
Thanks a lot!
bors r+

**Pull request merged**